### ISSUE TITLE
index out of range exception when using DELETE

### ISSUE THREAD

**Issue by mhingston** [cockroachdb/cockroach#59109]
Cockroach Core v20.2.3 (linux amd64)

For demonstration purposes I'm running a 3 node cluster (insecure) as described [here](https://www.cockroachlabs.com/docs/v20.2/start-a-local-cluster).

After completing step 1 above, once connected to the SQL client (node 1)...

```sql
CREATE DATABASE credits;
USE credits;

CREATE TABLE public.vehiclesummary (
	vehicle varchar(32) NOT NULL,
	vehicleid varchar(40) NOT NULL,
	vehiclename varchar(256),
	lastknowncredits double precision NOT NULL default 0.0,
	lastappearancedt timestamp NULL
);

CREATE UNIQUE INDEX vehiclesummary_vehicle_idx ON public.vehiclesummary (vehicle,vehicleid);
CREATE INDEX vehiclesummary_lastappearancedt_idx ON public.vehiclesummary (lastappearancedt DESC);
ALTER TABLE vehiclesummary ADD PRIMARY KEY (vehicle, vehicleid);

INSERT INTO public.vehiclesummary VALUES ('tag', 'AAAAAAAA', 'Foo', 0, '2020-01-01');

DELETE FROM public.vehiclesummary WHERE vehiclename = 'Foo';
```

Upon running the DELETE statement the node crashes...

```
*
* ERROR: [n1,client=127.0.0.1:43888,hostnossl,user=root] a SQL panic has occurred while executing the following statement:
* DELETE FROM public.vehiclesummary WHERE vehiclename = 'Foo'
*
*
* ERROR: [n1,client=127.0.0.1:43888,hostnossl,user=root] a panic has occurred!
* runtime error: index out of range [3] with length 3
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:506
*   | [...repeated from below...]
* Wraps: (2) while executing: DELETE FROM _._ WHERE _ = _
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:506
*   | runtime.gopanic
*   |   /usr/local/go/src/runtime/panic.go:679
*   | runtime.goPanicIndex
*   |   /usr/local/go/src/runtime/panic.go:75
*   | github.com/cockroachdb/cockroach/pkg/sql/rowenc.findColumnValue
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowenc/index_encoding.go:491
*   | github.com/cockroachdb/cockroach/pkg/sql/rowenc.EncodeColumns
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowenc/index_encoding.go:1677
*   | github.com/cockroachdb/cockroach/pkg/sql/rowenc.EncodeSecondaryIndex
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowenc/index_encoding.go:1001
*   | github.com/cockroachdb/cockroach/pkg/sql/row.(*Deleter).DeleteRow
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/row/deleter.go:104
*   | github.com/cockroachdb/cockroach/pkg/sql.(*tableDeleter).row
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/tablewriter_delete.go:55
*   | github.com/cockroachdb/cockroach/pkg/sql.(*deleteNode).processSourceRow
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/delete.go:176
*   | github.com/cockroachdb/cockroach/pkg/sql.(*deleteNode).BatchedNext
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/delete.go:116
*   | github.com/cockroachdb/cockroach/pkg/sql.(*rowCountNode).startExec
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_batch.go:159
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec.func2
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:514
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:119
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:298
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:86
*   | github.com/cockroachdb/cockroach/pkg/sql.walkPlan
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:50
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:517
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:125
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:774
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:392
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:422
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1002
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1001
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:872
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:639
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:114
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1465
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1467
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1391
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:508
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
*   |   /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:626
*   | runtime.goexit
*   |   /usr/local/go/src/runtime/asm_amd64.s:1357
* Wraps: (4) runtime error: index out of range [3] with length 3
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) runtime.boundsError
*
*
* ERROR: [n1,client=127.0.0.1:43888,hostnossl,user=root] Queued as error 3b052b66d12d43988689fe1d3edb2c5a
*
ERROR: driver: bad connection
warning: connection lost!
opening new connection: all session settings will be lost
warning: error retrieving the transaction status: dial tcp 127.0.0.1:26257: connect: connection refused
warning: connection lost!
opening new connection: all session settings will be lost
warning: error retrieving the database name: dial tcp 127.0.0.1:26257: connect: connection refused
```

**Comment by yuzefovich** [comment]
@mhingston thanks for the detailed report! On a quick glance I think you ran into #54629 which has been fixed on master branch, and the fix has been backported to 20.2 branch and will be included in the next point release (20.2.4) that is scheduled for January 19th.

**Comment by yuzefovich** [comment]
I'll close this as a dup of #54629. If you run into this problem after upgrading to 20.2.4, please file another report.
